packages/logging/src/logging-and-telemetry.ts

```diff
@@ -0,0 +1,785 @@
+/* eslint prefer-arrow-callback: "error" */
+import redactInfo from 'mongodb-redact';
+import { redactURICredentials } from '@mongosh/history';
+import type {
+  MongoshBus,
+  ApiEventWithArguments,
+  ApiEvent,
+  UseEvent,
+  EvaluateInputEvent,
+  ShowEvent,
+  ConnectEvent,
+  ScriptLoadFileEvent,
+  StartLoadingCliScriptsEvent,
+  StartMongoshReplEvent,
+  GlobalConfigFileLoadEvent,
+  CryptLibrarySkipEvent,
+  CryptLibraryFoundEvent,
+  SnippetsCommandEvent,
+  SnippetsErrorEvent,
+  SnippetsFetchIndexErrorEvent,
+  SnippetsFetchIndexEvent,
+  SnippetsLoadedEvent,
+  SnippetsLoadSnippetEvent,
+  SnippetsNpmDownloadActiveEvent,
+  SnippetsNpmDownloadFailedEvent,
+  SnippetsNpmLookupEvent,
+  SnippetsRunNpmEvent,
+  SnippetsTransformErrorEvent,
+  EditorRunEditCommandEvent,
+  EditorReadVscodeExtensionsDoneEvent,
+  EditorReadVscodeExtensionsFailedEvent,
+  FetchingUpdateMetadataEvent,
+  FetchingUpdateMetadataCompleteEvent,
+  SessionStartedEvent,
+  MongoshBusEventsMap,
+} from '@mongosh/types';
+import { inspect } from 'util';
+import { MongoLogWriter } from 'mongodb-log-writer';
+import { mongoLogId } from 'mongodb-log-writer';
+import type {
+  AnalyticsIdentifyMessage,
+  MongoshAnalytics,
+  MongoshAnalyticsIdentity,
+} from './analytics-helpers';
+import { hookLogger } from '@mongodb-js/devtools-connect';
+import { MultiSet, toSnakeCase } from './helpers';
+import { Writable } from 'stream';
+
+export class MongoshLoggingAndTelemetry {
+  log: MongoLogWriter;
+  pendingLogEvents: CallableFunction[] = [];
+
+  telemetryAnonymousId: string | undefined;
+  userId: string | undefined;
+  trackingProperties: {
+    mongosh_version: string;
+    session_id: string;
+  };
+  userTraits: AnalyticsIdentifyMessage['traits'] & {
+    [key: string]: unknown;
+  };
+
+  isSetup = false;
+  hasMongoLogWriterInitialized = false;
+
+  constructor(
+    public bus: MongoshBus,
+    public analytics: MongoshAnalytics,
+    public providedTraits: { platform: string } & {
+      [key: string]: unknown;
+    },
+    mongoshVersion: string
+  ) {
+    const dummySink = new Writable({
+      write(chunk, encoding, callback) {
+        callback();
+      },
+    });
+    this.log = new MongoLogWriter('', null, dummySink);
+
+    this.userTraits = {
+      ...providedTraits,
+      session_id: this.log.logId,
+    };
+
+    this.trackingProperties = {
+      mongosh_version: mongoshVersion,
+      session_id: this.log.logId,
+    };
+  }
+
+  public setup() {
```

Comment:
Is there any case where this isn't called immediately after the constructor?

Reply:
Yeah, I have had people prefer a constructor not to have "side effects" so wasn't sure about combining it. Though I myself don't see a problem, at least in this case.

Reply:
Yeah, fair enough, I didn't really think of setting up listeners as side effects here 🙂